### ISSUE TITLE
Loss function (summing up instead of averaging)

### ISSUE THREAD

**Issue by hurjunhwa** [ClementPinard/FlowNetPytorch#17]
Hi Clément,

I just noticed that you are working on refactoring the source code!
Thanks a lot!

Btw, I am currently using a modified version of your previous implementation, and I found that a simple modification on the loss design produced better results.

In your loss implementation (multiscaleloss.py), 
it first calculates the L2 distance (EPE) between the GT and output, and then it averages them over the image.
However, when just **"summing up the EPE"**, the network performs roughly similar to the original implementation.
(Actually I got near 2.2 EPE in the FlyingChairs from this modification + some additional minor things.)

I haven't thoroughly checked the original FlowNetS implementation in Caffe, but when looking at the scale of the loss function of theirs, I thought that summing up the L2 loss over the image seems the way that the original implementation took.

Could you check whether this is the case?

Thanks,
Jun

**Comment by ClementPinard** [comment]
Helo, thanks ! What do you mean by summing up the EPE instead of averaging it ? Do you sum every EPE ? Isn't it resulting in a gigantic loss ?

**Comment by hurjunhwa** [comment]
Yes exactly.
I am not sure whether I can understand your current version, 
but trying this **EPE_map.sum()** can be what I meant.

Yes it results a bit of gigantic loss.
The thing is that when I ran the original Caffe implementation, their loss was also quite huge as well.

**Comment by hurjunhwa** [comment]
Oops, **EPE_map.sum()** is the wrong way to do in the current version.
Please forget about it..

What I did in the previous one is following:

def EPE(input_flow, target_flow):
    return torch.norm(target_flow - input_flow, 2, 1)

    def forward(self, input, target, mode=0):
        if mode == 0:
            input_f = input[0]
            out = 0
            for i, input_ in enumerate(input_f):
                target_ = self.multiScales[i](target[0])
                for n in range(0, self.n_iter):
                    EPE_curr = EPE(input_[n], target_)
                    # Compare EPE_ with A Variable of the same size, filled with zeros)
                    # out += self.weights[i] * self.loss(EPE_curr, EPE_curr.detach() * 0) 
                    **out += self.weights[i] * EPE_curr.sum()**

**Comment by ClementPinard** [comment]
Well, in the previous version of the code, EPE_curr was a map, it's the same thing as now's EPE_map. I just launched a training using sum instead of mean, we'll see where it goes.

But essentially, summing instead of averaging is multiply every weight by the number of pixels, which means flow2 will be much more important. And it is consistant with the fact that I observed reversing the weights (more weight on flow2) conducted to better results.

**Comment by hurjunhwa** [comment]
Oh right.. yes, I have the same observation.
Intuitively, the larger scale map should get more weight, but the weights from Caffe implemention was not the case.

However, when summing up EPE, it just naturally becomes to put more weights for the higher maps.
So the relative scale becomes (intuitively speaking):
from (0.005, 0.01, 0.02, 0.08, 0.32)
to (1.28, 0.64, 0.32, 0.32, 0.32)..

**Comment by ClementPinard** [comment]
> Intuitively, the larger scale map should get more weight
Well, one could say that high res flow can only be fine-grained from lower resolutions flow fields, meaning that flow6 is top priority, and so on until flow2.

Anyway thanks for bringing it up ! 2.2 EPE seems sweet :) I'll be able to release networks trained from scratch that way !

**Comment by hurjunhwa** [comment]
Right, flow6 can be more important. My comment on it can be a bit controversial.. 

Yeap! hopefully it can make your implementation work better as well!
If it's not the case, then I will let you know what I changed from your previous implementation.
It should be one of them... :)

**Comment by ClementPinard** [comment]
Summing seems to do the trick, updated code and pretrained networks to follow ! Thanks again for your help ;)

**Comment by chenchr** [comment]
It seems that the caffe implementation is indeed not doing average. From the caffe proto for  L1LossParameter : ''l1loss_layer.cu', in 'the forward_gpu function:
```
message L1LossParameter {
  optional bool l2_per_location = 1 [default = false];
  optional bool l2_prescale_by_channels = 2 [default = false]; // Old style
  optional bool normalize_by_num_entries = 3 [default = false]; // if we want to normalize not by batch size, but by the number of non-NaN entries
  optional float epsilon = 4 [default = 1e-2]; // constant for smoothing near zero
  optional float plateau = 3001 [default = 0]; // L1 Errors smaller than plateau-value will result in zero loss and no gradient
}
```
there is a default setup '''normalize_by_num_entries''' is set to false, which means the loss is normalized by batch_size instead of num of element.
In the train.prototxt we can also see that:
```
layer {
  name: "flow_loss4"
  type: "L1Loss"
  bottom: "predict_flow4"
  bottom: "blob34"
  top: "flow_loss4"
  loss_weight: 0.02
  l1_loss_param {
    l2_per_location: true
  }
}
```
which use the default setup for ```normalize_by_num_entries```.
Them in  ''l1loss_layer.cu', in 'the forward_gpu function:
```
template <typename Dtype>
void L1LossLayer<Dtype>::Forward_gpu(const vector<Blob<Dtype>*>& bottom,
      const vector<Blob<Dtype>*>& top)
{
  
  Blob<Dtype> *diffptr = diff_top_vec_[0];
  
  Dtype dot, loss;
  if(bottom.size() > 1) {
    diff_layer_->Forward(bottom, diff_top_vec_);
  }
  
  // if necessary, compute the number of not-NaNs
  int count = bottom[0]->count();
  int num = bottom[0]->num();
  FindNotNaNs<Dtype><<<CAFFE_GET_BLOCKS(count), CAFFE_CUDA_NUM_THREADS>>>(
        count, diffptr->gpu_data(), mask_.mutable_gpu_data());
  cudaDeviceSynchronize();
  CUDA_POST_KERNEL_CHECK;
  
  if (this->layer_param_.l1_loss_param().normalize_by_num_entries()) {    
    caffe_gpu_dot(count, mask_.gpu_data(), mask_.gpu_data(), &normalize_coeff_);
    normalize_coeff_ /= mask_.channels();
  } else {
    normalize_coeff_ = num;
  }
  
  if (this->layer_param_.l1_loss_param().l2_per_location()) {
    // set masked (NaNs only) to zero
    KillMasked<Dtype><<<CAFFE_GET_BLOCKS(count), CAFFE_CUDA_NUM_THREADS>>>(
          count, mask_.gpu_data(), diffptr->mutable_gpu_data());
    cudaDeviceSynchronize();
    CUDA_POST_KERNEL_CHECK;
    
    square_layer_->Forward(diff_top_vec_, square_top_vec_);
    sum_layer_->Forward(square_top_vec_, sum_top_vec_);
    
    // Mask plateau in summed blob (only one channel):
    if(this->layer_param_.l1_loss_param().plateau() > 0) {
      float plateau_val_squared = this->layer_param_.l1_loss_param().plateau() * this->layer_param_.l1_loss_param().plateau();
      MaskPlateauValuesInitial<Dtype><<<CAFFE_GET_BLOCKS(sum_output_.count()), CAFFE_CUDA_NUM_THREADS>>>(
          sum_output_.count(), sum_output_.gpu_data(), plateau_l2_.mutable_gpu_data(), plateau_val_squared);
      cudaDeviceSynchronize();
      CUDA_POST_KERNEL_CHECK;
      
      KillMasked<Dtype><<<CAFFE_GET_BLOCKS(sum_output_.count()), CAFFE_CUDA_NUM_THREADS>>>(
            sum_output_.count(), plateau_l2_.gpu_data(), sum_output_.mutable_gpu_data());
      cudaDeviceSynchronize();
      CUDA_POST_KERNEL_CHECK;
    }
    
    sqrt_layer_->Forward(sum_top_vec_, sqrt_top_vec_);
    // Note sign_ is set to all ones in Reshape
    caffe_gpu_dot(sqrt_output_.count(), sqrt_output_.gpu_data(), sign_.gpu_data(), &dot);
  }
  else {    
    // Mask plateau:
    if(this->layer_param_.l1_loss_param().plateau() > 0) {
      MaskPlateauValues<Dtype><<<CAFFE_GET_BLOCKS(count), CAFFE_CUDA_NUM_THREADS>>>(
          count, diffptr->gpu_data(), mask_.mutable_gpu_data(), this->layer_param_.l1_loss_param().plateau());
      CUDA_POST_KERNEL_CHECK;
    }
    
    //mask_.print("MASK2");
    
    // set masked (NaNs, plateau) to zero
    KillMasked<Dtype><<<CAFFE_GET_BLOCKS(count), CAFFE_CUDA_NUM_THREADS>>>(
          count, mask_.gpu_data(), diffptr->mutable_gpu_data());
    CUDA_POST_KERNEL_CHECK;
    
    ComputeSign<Dtype><<<CAFFE_GET_BLOCKS(count), CAFFE_CUDA_NUM_THREADS>>>(
        count, diffptr->gpu_data(), sign_.mutable_gpu_data());
    CUDA_POST_KERNEL_CHECK;
    caffe_gpu_dot(count, diffptr->gpu_data(), sign_.gpu_data(), &dot); 
  }
  loss = dot / normalize_coeff_; 
  top[0]->mutable_cpu_data()[0] = loss;
}
```
we can see that firstly the code calculate the L2 norm and them sum up, and finally divides it by normalize_coeff_ in the code 
```loss = dot / normalize_coeff_; ``` .
Go back in the function we can see that the vatiable ```normalize_coeff_``` is determined in:
```
if (this->layer_param_.l1_loss_param().normalize_by_num_entries()) {    
    caffe_gpu_dot(count, mask_.gpu_data(), mask_.gpu_data(), &normalize_coeff_);
    normalize_coeff_ /= mask_.channels();
  } else {
    normalize_coeff_ = num;
  }
```
So, ```normalize_coeff_ = num;``` which is the batch size.
Therefore, maybe not just sum up, but need to divide it by the batch size..

**Comment by ClementPinard** [comment]
I also took a glance at the layer, and saw that summing was done, but I didn't see they divided by number of channels ! I will try it when I have, but I went to an EPE of 2.6 which is fine by now. Thanks a lot for your help anyway !

**Comment by chenchr** [comment]
I just tried to modify the code in ```multiscaleloss.py```. I have a question about the sparse situation, in the code:
```
def EPE(input_flow, target_flow, sparse=False):
    EPE_map = torch.norm(target_flow-input_flow,2,1)
    if sparse:
        return EPE_map[target_flow == 0].mean()
    else:
        return EPE_map.mean()
```
why only take the value at location that "target_flow == 0", in my naive understanding, it should take the value at location that "target_flow != 0"?

**Comment by ClementPinard** [comment]
You are right, i just pushed a correction along with some other updates, check it out ;)

**Comment by chenchr** [comment]
Thanks.

**Comment by chenchr** [comment]
Besides, it seems that the multi step lr implementation is not that right.
In caffe implemetation, lr is tuned in every iter, however in this pytorch implementation lr is tuned in every epoch.
Therefore, as in caffe:
```
net: "../model/train.prototxt"
test_iter: 640
test_interval: 5000
base_lr: 0.0001
lr_policy: "multistep"
gamma: 0.5
stepvalue: 300000
stepvalue: 400000
stepvalue: 500000
momentum: 0.9
weight_decay: 0.0004
display: 10
max_iter: 600000
snapshot: 20000
snapshot_prefix: "flow"
solver_mode: GPU
solver_type: ADAM
momentum2: 0.999
```
As by setting different batchsize, the train_loader will have different length.
I think we should divide the step, i.e. 300000, 400000 and 500000 by the length of train_loader to get the  step value in pytorch, which changes lr in a uint of epoch. Or we can adjust the lr in every train iteration instead of every epoch.

**Comment by hurjunhwa** [comment]
Yeah, true, but I found that slightly different training schedules don't really affect the results in the end.

I just use this setting with the batch size of 8:

    if epoch == 108 or epoch == 144 or epoch == 180:
        for param_group in optimizer.param_groups:
            param_group['lr'] = param_group['lr'] / 2

**Comment by ClementPinard** [comment]
Well 100K, 150K and 200K iterations seem pretty hardcore for milestones. I'm going to make them parametrable, but I prefer it to be related to the number of epochs.
Default parameters will match original hyper parameters, if you don't touch epoch size and training milestones

**Comment by chenchr** [comment]
@hurjunhwa hello. I found that in the flownet paper the test epe on chairs of flownets is 2.7, can you share how to achieve 2.2 epe? Thanks!

**Comment by ClementPinard** [comment]
Quick response before an achieved convergence :
![progress](https://i.imgur.com/gvod5Y3.png)

With first set of hyperparameters, EPE is down to 2.6 on test set
 - batch size of 16
 - epochs of 500 batches
 - lr 1e-4
 - default data augmentation, including color warp

With second set of hyper parameters, training is still ongoing, but EPE is already down to 2.35
 - batch size of 8
 - epochs of 1000 batches
 - lr 1e-4
 - default data augmentation, apart from color warping, which is shut down

**Comment by hurjunhwa** [comment]
Awesome!!

@chenchr : here is the list that I changed.
Some of them may be different from the original implementation because I didn't fully check the original source code.
This modification is based on the previous version of Clement's implementation.

- Loss.
 As explained above, this was the most critical one (summing up the EPE instead of averaging).
 However, as you suggested, it should be divided by the batch size.

- Photometric augmentation
 I did multiplicative, additive brightness changes, and contrast change.
When doing this, it seems to generalize the result better (closing the gap between train EPE & validation EPE).

- Geometric augmentation
 For random rotation, I used this coefficient:
  flow_transforms.RandomRotate(20, 3),
 When using the original one "flow_transforms.RandomRotate(10, 5)", it created some minor artifacts on the augmented image.
Thus, reduce the value of the second parameter.

- Input scale
 I scale the input image to [0, 1].
 I didn't normalize it further.. (I am not so sure whether it's okay not to do it..)

- EPE calculation
 I always resize the output flow to the same as that of original image, and calculate the EPE.

- Architecture
 I didn't use the batch normalization.
 In FlowNetS, I turned on the bias term for predict_flow (nn.Conv2d) & upsample_flow layers (nn.ConvTranspose2D).
However I am not so sure whether this is the right way to do (when looking at the Caffe version, it seems that they have the bias term for these layers).
 I used the same initialization as the Caffe version.

- Training schedule
 Following the same setting in Caffe version.
(lr, batch size 8, adam parameters, weight decay, etc..)


I didn't do the ablation study, so I am not sure which one is effective & which one is not necessary.
Also, some of them could be a wrong way to do it.

In my opinion, the loss function & the way augmentation would be the two critical things that majorly affects the results in the end.

**Comment by ClementPinard** [comment]
> When doing this, it seems to generalize the result better (closing the gap between train EPE & validation EPE).

The gap is actually in the right order ! :) test EPE is already lower than train EPE, so you actually want to expand it ! This is mainly because of the picture size in train set (they are cropped to `[320,448]`)

**Comment by chenchr** [comment]
@hurjunhwa Thanks for your elaborate explaination.

**Comment by hurjunhwa** [comment]
You are welcome!
I guess there is no golden standard setting, and thus it's okay to be different from the original one..
As long as the results are reasonable, it seems okay to keep the model as a baseline for your experiment or publication!

**Comment by chenchr** [comment]
Yes. I am just curious about that the epe 2.3 or 2.2 is smaller than the one reported in original paper. Besides, I just realize that you are the author of the paper unsupervised learning of optical flow from your page. It's a awesome work!

**Comment by hurjunhwa** [comment]
Thanks for the complement!

**Comment by chenchr** [comment]
@hurjunhwa Sorry to disturb you again. Have you met a situation that the test loss is lower than train loss? When I turn off the data augmentation, the test loss is a little bit higher than train loss, which is rational. However, when I turn on the data augmentation, the test loss is always lower than train loss. It's weird..

**Comment by ClementPinard** [comment]
> test EPE is already lower than train EPE, so you actually want to expand it ! This is mainly because of the picture size in train set (they are cropped to [320,448])

Thus, every point has a lower mean receptive field during traing than during test since pictures are smaller. To convince yourself you can try a trained network with data augmentation, and then run an inference on train set, you'll likely have very low EPE

**Comment by chenchr** [comment]
On, I got your point. I really appreciate your help as this phenomenon did make me confused these day.

**Comment by ClementPinard** [comment]
Training is now finished.
EPE went down to 1.951 !!!
https://drive.google.com/open?id=1jbWiY1C_nqAUJRYZu7mwzV6CK7ugsa5v

This is even better than FlowNetC by a large margin (reported EPE was ~2.1)

**Comment by chenchr** [comment]
@ClementPinard  Thank you for sharing the result. I found in the newest code https://github.com/ClementPinard/FlowNetPytorch/blob/master/multiscaleloss.py#L12
the line:
```
def EPE(input_flow, target_flow, sparse=False, mean=True):
    EPE_map = torch.norm(target_flow-input_flow,2,1)
    if sparse:
        EPE_map = EPE_map[target_flow != 0]
    if mean:
        return EPE_map.mean()
    else:
        return EPE_map.sum()
```
I think if not divide the epe by batchsize, then the learning rate for different batchsize setup will be different during training.
